"panic: invalid union tag" during collect()

Repro is with Zed commit 70e7129. This is an anonymized example of a crash that was found when debugging a `join` from a community zync user.

Here's sample data `anon.json` that reproduces the problem:

```
{
  "key": {
    "c_id": 6,
    "t_id": 3
  },
  "value": {
    "before": {
      "c_id": 6,
      "t_id": 3
    }
  }
}
{
  "key": {
    "c_id": 6,
    "t_id": 3
  },
  "value": null
}
```

Prepping it in a pool:

```
$ zed -version
Version: v1.8.1-61-g70e71293

$ rm -rf lake

$ zed -lake lake init
lake created: lake

$ zed -lake lake create anon
pool created: anon 2SG7zDMTnFWaRqCk3xfIoJ2eT6j

$ zed -lake lake load -use anon anon.json
(2/1) 186B/186B 186B/s 100.00%
2SG80vqF124Gq09RuMD3Ur8Pg53 committed
```

Now to trigger the crash:

```
$ zed -lake lake query 'from anon | put extrakey := key | c := collect(this) by extrakey'
panic: invalid union tag

goroutine 1 [running]:
github.com/brimdata/zed.(*TypeUnion).Untag(0xc000808120, {0xc0000ea9f7?, 0x30?, 0x24874a0?})
	/Users/phil/work/zed/complex.go:360 +0x125
github.com/brimdata/zed/zson.(*elemHelper).add(0xc00098f748, {0xc0000ea9f7?, 0x0?, 0x0?})
	/Users/phil/work/zed/zson/formatter.go:476 +0x45
github.com/brimdata/zed/zson.(*Formatter).formatVector(0xc000796510, 0x0, {0x2781be8?, 0x100000000000c?}, {0x2781c18, 0x1}, {0x279ad98, 0xc000808120}, 0xc00098f878, 0x0, ...)
	/Users/phil/work/zed/zson/formatter.go:440 +0x3fb
github.com/brimdata/zed/zson.(*Formatter).formatValue(0xc000796510, 0x3303718?, {0x279a960?, 0xc000810078?}, {0xc0000ea9f6, 0x28, 0x2a}, 0x0, 0x0, 0x1)
	/Users/phil/work/zed/zson/formatter.go:173 +0x316
github.com/brimdata/zed/zson.(*Formatter).formatRecord(0xc000796510, 0x0?, 0xc000808210?, {0xc0000ea9f0?, 0x2e?, 0x30?}, 0x80?, 0x33?)
	/Users/phil/work/zed/zson/formatter.go:416 +0x2e5
github.com/brimdata/zed/zson.(*Formatter).formatValue(0xc000796510, 0xc000808210?, {0x279ad48?, 0xc000808210?}, {0xc0000ea9f0, 0x2e, 0x30}, 0x0, 0x0, 0x0)
	/Users/phil/work/zed/zson/formatter.go:171 +0x27e
github.com/brimdata/zed/zson.(*Formatter).formatValueAndDecorate(0xc000796510, {0x279ad48?, 0xc000808210?}, {0xc0000ea9f0, 0x2e, 0x30})
	/Users/phil/work/zed/zson/formatter.go:146 +0x15c
github.com/brimdata/zed/zson.(*Formatter).FormatRecord(0xc000796510, 0xc0008cda70)
	/Users/phil/work/zed/zson/formatter.go:75 +0xb9
github.com/brimdata/zed/zio/zsonio.(*Writer).Write(0xc000746810, 0xc0008cc700?)
	/Users/phil/work/zed/zio/zsonio/writer.go:33 +0x29
github.com/brimdata/zed/zio.CopyWithContext({0x27a57c8, 0xc0000be018}, {0x2c2f7740, 0xc000746810}, {0x2791be0, 0xc0008baaf0})
	/Users/phil/work/zed/zio/zio.go:158 +0x7f
github.com/brimdata/zed/zio.Copy(...)
	/Users/phil/work/zed/zio/zio.go:146
github.com/brimdata/zed/cmd/zed/query.(*Command).Run(0xc00027f040, {0xc0000ce040, 0x1, 0x0?})
	/Users/phil/work/zed/cmd/zed/query/command.go:72 +0x385
github.com/brimdata/zed/pkg/charm.path.run({0xc000266590?, 0x2, 0x2}, {0xc0000ce040?, 0x1, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x8b
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0xc0005b2f80?, {0xc0000ce010, 0x4, 0x4})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x4a
main.main()
	/Users/phil/work/zed/cmd/zed/main.go:66 +0x1125
```

Oddly, I can't reproduce the problem when using `zq` on the original file. 🤔 

```
$ zq -z 'file anon.json | put extrakey := key | c := collect(this) by extrakey'
{extrakey:{c_id:6,t_id:3},c:[{key:{c_id:6,t_id:3},value:{before:{c_id:6,t_id:3}},extrakey:{c_id:6,t_id:3}},{key:{c_id:6,t_id:3},value:null,extrakey:{c_id:6,t_id:3}}]}
```

--- a/runtime/expr/agg/collect.go
+++ b/runtime/expr/agg/collect.go
@@ -22,11 +22,7 @@ func (c *Collect) Consume(val *zed.Value) {
 }
 
 func (c *Collect) update(val *zed.Value) {
-	// Untag the value if it's a union.
-	if union, ok := zed.TypeUnder(val.Type).(*zed.TypeUnion); ok {
-		val = zed.NewValue(union.Untag(val.Bytes()))
-	}
-	c.values = append(c.values, *val.Copy())
+	c.values = append(c.values, *val.Under(&zed.Value{}).Copy())
 	c.size += len(val.Bytes())
 	for c.size > MaxValueSize {
 		// XXX See issue #1813.  For now we silently discard entries

--- a/runtime/expr/agg/collect.go
+++ b/runtime/expr/agg/collect.go
@@ -22,6 +22,10 @@ func (c *Collect) Consume(val *zed.Value) {
 }
 
 func (c *Collect) update(val *zed.Value) {
+	// Untag the value if it's a union.
+	if union, ok := zed.TypeUnder(val.Type).(*zed.TypeUnion); ok {
+		val = zed.NewValue(union.Untag(val.Bytes()))
+	}
 	c.values = append(c.values, *val.Copy())
 	c.size += len(val.Bytes())
 	for c.size > MaxValueSize {

--- a/runtime/expr/agg/ztests/collect-untag-unions.yaml
+++ b/runtime/expr/agg/ztests/collect-untag-unions.yaml
@@ -3,7 +3,7 @@ zed: collect(this)
 input: |
   1((int64,string))
   "foo"((time,string))
-  .5
+  0.5
 
 output: |
   [1,"foo",0.5]

--- a/runtime/expr/agg/ztests/collect-untag-unions.yaml
+++ b/runtime/expr/agg/ztests/collect-untag-unions.yaml
@@ -1,0 +1,9 @@
+zed: collect(this)
+
+input: |
+  1((int64,string))
+  "foo"((time,string))
+  .5
+
+output: |
+  [1,"foo",0.5]

